doc/users-guide/users-guide-ipsec.adoc

```diff
@@ -0,0 +1,443 @@
+facilities that ODP provides. However, since especially with inline IPsec
+processing, the application may not have explicit knowledge of the traffic
+volumes associated with a given SA, support for usage-based limits is
+integrated into ODP IPsec support.
+
+At `odp_ipsec_sa_create()` time, one of the fields in the
+`odp_ipsec_sa_param_t` struct is the `odp_ipsec_lifetime_t` sub-structure.
+This struct allows hard and/or soft limits to be specified in terms of total
+bytes encrypted/decrypted, total packet count, or both. A limit specification
+of 0 indicates no limit for that metric. If either is specified, the limit
+is triggered on whichever occurs first. Given the defined behavior of hard vs.
+soft limits, the soft limits, if used, should always be specified as lower
+than the hard limits. These should be sufficiently lower to enable adequate
+time to switch over to a replacement SA before the hard limit is reached.
+
+As noted, when an SA hard limit is reached the SA immediately enters the
+expired state and attempts to use it further are failed with an
+`odp_ipsec_result_t` that indicates a hard expiration limit. When a soft
+limit is reached for packets sent via `odp_ipsec_out_inline()`, this results
+in an `ODP_EVENT_IPSEC_STATUS` event being sent to the application on the
+queue associated with the SA that has reached the soft limit. This status
+event has an `odp_ipsec_status_id_t` of `ODP_IPSEC_STATUS_WARN` with a
+`odp_ipsec_warn_t` bits set to indicate the type of soft expiration reached.
+Receipt of this event alerts the application that the SA is nearing the end of
+its useful life and that it should be replaced. It is the application's
+responsibility to heed this warning. Soft limit alerts are edge triggered and
+are only issued once rather than repeatedly as the SA runs beyond the soft
+limit.
+
+When operating in lookaside mode, expiration limits are carried as a warning
+in the `odp_op_status_t` section of the `odp_ipsec_result_t` struct. The same
+is true for inline inbound packets. When the soft limit is reached, these
+packets will carry a warning flag indicating this condition.
+
+==== SA Disablement and Destruction
+When it is time to retire an SA, the application does so by issuing a call
+to the `odp_ipsec_sa_destroy()` API. However, to better support graceful SA
+termination processing, ODP also provides the `odp_ipsec_sa_disable()` API.
+
+This call initiates termination processing for an SA by stopping use of the SA
+for new operations while still allowing those that are "in flight" to complete
+processing. A special "dummy" packet with an `odp_ipsec_result_t` that says
+this SA is now disabled indicates that the packet pipeline for this SA is now
+clear. Note that these dummy packets will have an `odp_packet_len()` of zero
+and should not be used as actual packet data. Upon receipt of such dummy
+packets the application should simply take appropriate action and then free
+the event.
```

Comment:
This dummy packet stuff is not yet (if ever?) part of the API spec. I think this patch should be in sync with what is the current state of api-next, which still uses the status event for indicating SA disable completion.

Reply:
This version of the PR was intended to match the changes in PR #256. I'll revert these in the next version since it seems we're now content with the original semantics.